diffrax/_progress_bar.py

```diff
@@ -0,0 +1,186 @@
+
+
+class AbstractProgressBar(eqx.Module, Generic[ProgressBarGenericState]):
+    @abc.abstractmethod
+    def init(self) -> ProgressBarGenericState:
+        """Creates a new progress bar and returns its unique identifier"""
+
+    @abc.abstractmethod
+    def step(
+        self, state: ProgressBarGenericState, progress: FloatScalarLike
+    ) -> ProgressBarGenericState:
+        """Updates the progress bar and returns the last printed progress value"""
+
+    @abc.abstractmethod
+    def close(self, state: ProgressBarGenericState):
+        """Closes the progress bar"""
+
+
+NoProgressBarState = NamedTuple("NoProgressBarState", [])
+
+
+class NoProgressBar(AbstractProgressBar):
+    def init(self) -> NoProgressBarState:
+        return NoProgressBarState()
+
+    def step(
+        self,
+        state: NoProgressBarState,
+        progress: FloatScalarLike,
+    ) -> NoProgressBarState:
+        return state
+
+    def close(self, state: NoProgressBarState):
+        pass
+
+
+_TextProgressBarState = NamedTuple(
+    "_TextProgressBarState", [("progress", FloatScalarLike)]
+)
+
+
+class TextProgressBar(AbstractProgressBar):
+    minimum_increase: float
+
+    def __init__(self, minimum_increase: float = 2.0):
+        """
+        **Arguments:**
+            - `minimum_increase`: minimum increase of progress bar in percent
+        """
+        self.minimum_increase = minimum_increase
+
+    def init(self) -> _TextProgressBarState:
+        return _TextProgressBarState(progress=0.0)
+
+    def step(
+        self,
+        state: _TextProgressBarState,
+        progress: FloatScalarLike,
+    ) -> _TextProgressBarState:
+        """
+        **Arguments:**
+            - `state`: progress bar state
+            - `progress`: current progress in percent
+        """
+
+        def update_aux(progress):
+            # currently
+            # jax.debug.print("{progress:.2f}%", progress=progress)
+            # does not work, so we have to round the progress ourselves
+            jax.debug.print("{}%", np.round(progress, 1), ordered=True)
+            return progress
+
+        # we only print if the progress has increased by at least
+        # `minimum_increase` to avoid flooding the user
+        # with too many updates
+        next_progress = jax.lax.cond(
+            progress - state.progress > self.minimum_increase,
+            update_aux,
+            lambda p: state.progress,
+            progress,
+        )
+
+        return _TextProgressBarState(progress=next_progress)
+
+    def close(self, state: _TextProgressBarState):
+        jax.debug.print("100%", ordered=True)
+
+
+_TqdmProgressBarState = NamedTuple(
+    "_TqdmProgressBarState", [("pb_id", IntScalarLike), ("step", int)]
+)
+
+
+class TqdmProgressBar(AbstractProgressBar):
+    refresh_rate: int
+
+    def __init__(self, refresh_rate: int = 20):
+        """
+        **Arguments:**
+            - `refresh_rate`: number of steps between refreshes
```

Comment:
Nit: this won't format as Markdown correctly; it needs a newline between the "Arguments" and the list, and the list should be deindented. (Take a look at some of the existing code for an example.)

This matters as we'll put this in the documentation, so it has to parse correctly.